corehq/apps/users/audit/change_messages.py

```diff
@@ -1,91 +1,329 @@
-class UserChangeMessage(object):
+from enum import Enum
+from django.utils.translation import ugettext as _
+
+
+class Change(Enum):
+    SET = 'set'
+    RESET = 'reset'
+    REMOVE = 'remove'
+    ADD = 'add'
+
+
+class UserChangeMessageV1(object):
+    """
+    Each change message to follow the structure
+    {
+        "field": {
+            "change": change_details
+        }
+    }
+    field: could be domain, phone_numbers etc
+    change: either 'set' / 'reset' / 'remove' / 'add'
+    change_details: a Boolean or
+                    a dict with details as needed for the change
+                    a list of a dicts as needed for the change
+    """
     @staticmethod
     def program_change(program):
         if program:
-            message = f"Program: {program.name}[{program.get_id}]"
+            change_message = {"program": {Change.SET: {"id": program.get_id, "name": program.name}}}
         else:
-            message = 'Program: None'
-        return message
+            change_message = {"program": {Change.SET: {}}}
+        return change_message
 
     @staticmethod
     def role_change(user_role):
-        message = 'Role: None'
         if user_role:
-            message = f"Role: {user_role.name}[{user_role.get_qualified_id()}]"
-        return message
+            change_message = {"role": {Change.SET: {"id": user_role.get_qualified_id(), "name": user_role.name}}}
+        else:
+            change_message = {"role": {Change.SET: {}}}
+        return change_message
 
     @staticmethod
     def domain_removal(domain):
-        return f"Removed from domain '{domain}'"
+        return {"domain": {Change.REMOVE: {"name": domain}}}
 
     @staticmethod
     def registered_devices_reset():
-        return "Registered devices reset"
+        return {"devices": {Change.RESET: True}}
 
     @staticmethod
     def two_factor_disabled_for_days(days):
-        return f"Disabled for {days} days"
+        return {"two_factor": {Change.REMOVE: {"days": days}}}
 
     @staticmethod
     def two_factor_disabled_with_verification(verified_by, verification_mode):
-        return f'Two factor disabled. Verified by: {verified_by}, verification mode: "{verification_mode}"'
+        return {
+            "two_factor": {
+                Change.REMOVE: {
+                    "verified_by": verified_by,
+                    "verification_mode": verification_mode
+                }
+            }
+        }
 
     @staticmethod
     def password_reset():
-        return "Password reset"
+        return {"password": {Change.RESET: True}}
 
     @staticmethod
-    def status_update(action, reason):
-        return f'User {action}. Reason: "{reason}"'
+    def status_update(active, reason):
+        return {
+            "status": {
+                Change.SET: {
+                    "active": active,
+                    "reason": reason
+                }
+            }
+        }
```

Comment:
Shouldn't this be:

```suggestion
        return {
            "is_active": {
                Change.SET: {
                    "value": active,
                    "reason": reason
                }
            }
        }
```

Also I see that this comes from an admin view that also logs a password reset. Should these not be combined since they are a single logical change?

Reply:
> Also I see that this comes from an admin view that also logs a password reset. Should these not be combined since they are a single logical change?

I will take a look at combining logs from the same view, that can be combined. Although a view would save all in one log only. So, say, if two-factor is being disabled and password is being reset in the same view, messages for these would be passed once in the end to log.

Plus,
It would be nice to have this class return messages for the smallest possible change and not combine them. Would make future versioning easy as well in case we need to change a single message.

Reply:
I think we should try and match the logs with the user action that was taken. So if a use changes their password then we just log that but if an admin user disables an account (and part of that is resetting the password) then we would just record that the account was disabled as one log and not separate logs for each field that was changed.

Reply:
> then we would just record that the account was disabled

Would that mean adding a new action to the list of "set"/"reset"/"add"/"remove" ?

> not separate logs for each field that was changed.

Just confirming that you mean separate messages in the same log? There wont be separate logs or User History records. Here is an example, https://github.com/dimagi/commcare-hq/pull/30181/files#diff-e0a7331ea888053cfc991192ff9122d31820052cf81b278b2cabc2aa8c95cab9R505